src/main/java/com/example/bookshop/service/impl/BookServiceImpl.java

```diff
@@ -21,12 +26,22 @@
 @Service
 public class BookServiceImpl implements BookService {
     private final BookRepository bookRepository;
+    private final CategoryRepository categoryRepository;
     private final SpecificationBuilder<Book> bookSpecificationBuilder;
     private final BookMapper bookMapper;
 
     @Override
     public BookDto save(CreateBookRequestDto bookRequestDto) {
         Book book = bookMapper.toBookEntity(bookRequestDto);
+
+        Set<Category> categories = bookRequestDto.getCategoryIds().stream()
```

Comment:
I think it will be better to just save the book without checking categories (it will increase performance because one request to db will be removed). But if you think that is necessary please create a method in categoryService to validate the category it will be better from a single responsibility principle point of view.